API Client Generation: Add documentation (#102056)
grafana/grafana: public/app/api/README.md

## Generating RTK Query API Clients

To show the steps to follow, we are going to work on adding an API client to create a new dashboard. Just adapt the following guide to your use case.

### 1. Generate an OpenAPI snapshot

First, check if the `group` and the `version` are already present in [openapi_test.go](/pkg/tests/apis/openapi_test.go). If so, move on to the next step.
<br/> If you need to add a new block, you can check for the right `group` and `version` in the backend API call that you want to replicate in the frontend.

```jsx
{
    Group:   "dashboard.grafana.app",

> 	Version: "v0alpha1",

Reviewer: nit: can we fix the indentation here 🙏

}
```

> ### 2. Run the `TestIntegrationOpenAPIs` test

Reviewer: imo this shouldn't be a separate step, it's part of the `Generate an OpenAPI snapshot` step

> Note that it will fail the first time you run it, and, in the second one, it will generate the corresponding OpenAPI spec. You will find it in [openapi_snapshots](/pkg/tests/apis/openapi_snapshots).

Reviewer 2: ```suggestion
Note that it will fail the first time you run it. On the second run, it will generate the corresponding OpenAPI spec, which you can find in [openapi_snapshots](/pkg/tests/apis/openapi_snapshots).
```

<br/>
<br/>

> > Note: You won't need to follow these two steps if the `group` you will work within is already in the `openapi_test.go` file.

Reviewer 2: ```suggestion
> Note: You don’t need to follow these two steps if the `group` you’re working with is already in the `openapi_test.go` file.
```

<br/>

### 3. Create the API definition

> In the folder `../public/app/features/{your_group_name}/api/` you have to create the `baseAPI.ts` file for your group. This file should have the following content:
> 

Reviewer 2: ```suggestion
In the `../public/app/features/{your_group_name}/api/` folder you have to create the `baseAPI.ts` file for your group. This file should have the following content:

```

| apiImport       | Function name exported in the API definition (baseAPI.ts file).                                                                                                                                                                                                                           |
| filterEndpoints | The `operationId` of the particular route you want to work with. You can check the available operationIds in the specific group's spec file. As seen in the `migrate-to-cloud` one, it is an array                                                                                        |
|  tag            | Must be set to `true`, to automatically attach tags to endpoints. This is needed for proper cache invalidation. See more info in the [official documentation](https://redux-toolkit.js.org/rtk-query/usage/automated-refetching#:~:text=RTK%20Query%20uses,an%20active%20subscription.).  |

> |  hooks          | Must be set to `false` since we only want to export the hooks from the index file, not the generated one.                                                                                                                                                                                 |

Reviewer: this defaults to false, so we can probably just remove all mention of it 😄

In our case, the dashboard index will be like:

```jsx
import { generatedAPI } from './endpoints.gen';

export const dashboardAPI = generatedAPI;
export const { useCreateDashboardMutation} = dashboardAPI;
// eslint-disable-next-line no-barrel-files/no-barrel-files
export { type Dashboard } from './endpoints.gen';

```

There are some use cases where the hook will not work, and that is a clue to see if it needs to be modified. The hooks can be tweaked by using `enhanceEndpoints`.

```jsx
export const dashboardsAPI = generatedApi.enhanceEndpoints({
  endpoints: {
    // Need to mutate the generated query to set the Content-Type header correctly
    createDashboard: (endpointDefinition) => {
      const originalQuery = endpointDefinition.query;
      if (originalQuery) {
        endpointDefinition.query = (requestOptions) => ({
          ...originalQuery(requestOptions),
          headers: {
            'Content-Type': 'application/merge-patch+json',
          },
        });
      }

>     },

Reviewer 2: This example is a bit weird, why do we use patch content type for a POST request? Did you maybe mean `updateDashboard`?

  },
});
```

> ### 7. Add reducers and middleware to the redux store

Reviewer 2: ```suggestion
### 7. Add reducers and middleware to the Redux store
```

> The last, but not least, step to be done is adding the middleware and reducers to the store.

Reviewer 2: ```suggestion
Last but not least, you need to add the middleware and reducers to the store.
```

In Grafana, the reducers are added to [`root.ts`](public/app/core/reducers/root.ts):

```jsx
  import { dashboardAPI } from '<pathToYourAPI>';
  const rootReducers = {
    ...,
    [dashboardAPI.reducerPath]: dashboardAPI.reducer,
  };
```

> And the middlewares are added to [`configureStore.ts`](public/app/store/configureStore.ts):

Reviewer 2: ```suggestion
And the middleware is added to [`configureStore.ts`](public/app/store/configureStore.ts):
```

```jsx
import { dashboardAPI } from '<pathToYourAPI>';
export function configureStore(initialState?: Partial<StoreState>) {
  const store = reduxConfigureStore({
    reducer: createRootReducer(),
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({ thunk: true, serializableCheck: false, immutableCheck: false }).concat(
        ...,
        dashboardAPI.middleware
      ),
    ...,
  });
```

> You have available the oficial documentation in [RTK Query](https://redux-toolkit.js.org/tutorials/rtk-query#add-the-service-to-your-store)

Reviewer: ```suggestion
You have available the official documentation in [RTK Query](https://redux-toolkit.js.org/tutorials/rtk-query#add-the-service-to-your-store)
```
